src/ml_board/backend/restful_api/data_access.py

```diff
@@ -236,6 +240,70 @@ def get_experiment_config(self, grid_search_id: str, experiment_id: str, config_
+        """
+        Fetch all checkpoint resource pickle files from event storage
+        given the epoch, experiment ID & grid search ID.
+
+        :params:
+             grid_search_id (str): Grid Search ID
+             experiment_id (str): Experiment ID
+             epoch (str): Epoch number
+
+        :returns: List of Checkpoint files
+        """
+        requested_full_path = os.path.realpath(
+            os.path.join(self.top_level_logging_path, str(grid_search_id), str(experiment_id), str(epoch))
+        )
+
+        if FileDataAccess.is_safe_path(base_dir=self.top_level_logging_path, requested_path=requested_full_path):
+            response = []
+            checkpoints = []
+            files = FileDataAccess.get_checkpoint_files(requested_full_path, base_path=self.top_level_logging_path)
+            for file_num in range(len(files)):
+                split = os.path.normpath(files[file_num]).split(os.sep)
+                checkpoints.append(os.path.basename(split[3]).split(".")[0])
+                if file_num == len(files) - 1:
+                    response.append({"experiment_id": experiment_id, "epoch": epoch, "checkpoints": checkpoints})
+            return response
+
+        else:
+            raise InvalidPathError(f"File path {requested_full_path} is not safe.")
+
+    def get_checkpoint_list(self, grid_search_id: str, experiment_id):
+        """
+        `Fetch checkpoint resource pickle file given the experiment ID & grid search ID from event storage.
+
+        :params:
+             grid_search_id (str): Grid Search ID
+             experiment_id (str): Experiment ID
+
+        :returns: list of checkpoints
+        """
+        requested_full_path = os.path.realpath(os.path.join(self.top_level_logging_path, str(grid_search_id), str(experiment_id)))
+
+        if FileDataAccess.is_safe_path(base_dir=self.top_level_logging_path, requested_path=requested_full_path):
+            response = []
+            checkpoints = []
+            files = FileDataAccess.get_checkpoint_files(requested_full_path, base_path=self.top_level_logging_path)
+            last_epoch = os.path.normpath(files[0]).split(os.sep)[2]
+            for file_num in range(len(files)):
+                split = os.path.normpath(files[file_num]).split(os.sep)
+                epoch = split[2]
+                if epoch != last_epoch:
+                    response.append({"experiment_id": experiment_id, "epoch": last_epoch, "checkpoints": checkpoints})
+                    last_epoch = epoch
+                    checkpoints = []
+                    checkpoints.append(os.path.basename(split[3]).split(".")[0])
+                else:
+                    checkpoints.append(os.path.basename(split[3]).split(".")[0])
+                    if file_num == len(files) - 1:
+                        response.append({"experiment_id": experiment_id, "epoch": epoch, "checkpoints": checkpoints})
```

Comment:
I think this assumes that files is always sorted. If this is not the case, then this code won't work anymore. --> code is prone to side-effects.

Reply:
the call: files = FileDataAccess.get_checkpoint_files(requested_full_path, base_path=self.top_level_logging_path) always returns an ordered file list since the files are always stored in an sorted manor. Practically it is not possible to have an unordered list but yes it is prone to side effects so i will add a sort function to be sure.